### PR TITLE
Add PolynomialCrossing to Addons

### PR DESCRIPTION
**Pull request by tanzhenyu**
This is from internal design to open-source the DCN network from paper https://arxiv.org/abs/1708.05123

### CODE REVIEW

**Comment by seanpmorgan** [comment]
Majority of issues seem to stem from formatting. Could you please run `black ./` or `bash tools/run_docker.sh -c 'make code-format'`

**Comment by tanzhenyu** [comment]
> Majority of issues seem to stem from formatting. Could you please run `black ./` or `bash tools/run_docker.sh -c 'make code-format'`

Done.

File: tensorflow_addons/layers/polynomial_test.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+@test_utils.run_all_in_graph_and_eager_modes
+class PolynomialCrossingTest(tf.test.TestCase):
+    # Do not use layer_test due to multiple inputs.
+
+    def test_full_matrix(self):
+        x0 = np.random.random((12, 5))
+        x = np.random.random((12, 5))
+        layer = PolynomialCrossing(projection_dim=None)
+        layer([x0, x])
```
Comment (seanpmorgan) [inline tensorflow_addons/layers/polynomial_test.py]: Please refactor with known input values and confirm the result with an assert. That way if anything changes we'll be notified in failing test.

File: tensorflow_addons/layers/polynomial.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        self.bias_initializer = tf.keras.initializers.get(bias_initializer)
+        self.kernel_regularizer = tf.keras.regularizers.get(kernel_regularizer)
+        self.bias_regularizer = tf.keras.regularizers.get(bias_regularizer)
+
+        self.supports_masking = True
+
+    def build(self, input_shape):
+        if not isinstance(input_shape, (tuple, list)) or len(input_shape) != 2:
+            raise ValueError(
+                "Input shapes must be a tuple or list of size 2, "
```
Comment (seanpmorgan) [inline tensorflow_addons/layers/polynomial.py]: Please add a test case for this error

File: tensorflow_addons/layers/polynomial.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+                "Inputs to the layer must be a tuple or list of size 2, "
+                "got {}".format(inputs)
+            )
+        x0, x = inputs
+        outputs = x0 * tf.matmul(x, self.kernel) + x
+        if self.use_bias:
+            outputs = tf.add(outputs, self.bias)
+        return outputs
+
+    def get_config(self):
```
Comment (seanpmorgan) [inline tensorflow_addons/layers/polynomial.py]: Please add a test for serialization. Similar to https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/tlu_test.py#L49

File: tensorflow_addons/layers/polynomial_test.py (hunk 1 above)
Comment (tanzhenyu) [reply tensorflow_addons/layers/polynomial_test.py]: Done.

File: tensorflow_addons/layers/polynomial.py (hunk 2 above)
Comment (tanzhenyu) [reply tensorflow_addons/layers/polynomial.py]: Done.

File: tensorflow_addons/layers/polynomial.py (hunk 3 above)
Comment (tanzhenyu) [reply tensorflow_addons/layers/polynomial.py]: Done.

**Comment by tanzhenyu** [comment]
@seanpmorgan Thanks Sean for the review! Any idea on what the sanity check is doing?
error message - "the remote end hung up unexpectedly; early EOF; index-pack failed; Git fetch failed with exit code: 128"
https://github.com/tensorflow/addons/pull/1018/checks?check_run_id=424954024

**Comment by gabrieldemarmiesse** [comment]
I think it's just a connection issue. I'll restart the tests.

**Comment by tanzhenyu** [comment]
> I think it's just a connection issue. I'll restart the tests.

Thanks!

**Comment by tanzhenyu** [comment]
> LGTM thanks for the PR!

Thanks for the review!

**Pull request merged**